core/src/main/java/org/mule/runtime/core/DefaultMuleEvent.java

```diff
@@ -990,6 +958,37 @@ public void setSecurityContext(SecurityContext context) {
     session.setSecurityContext(context);
   }
 
+  private MuleEvent parent;
+  private Correlation correlation;
+
+  public void setParent(MuleEvent parent) {
+    this.parent = parent;
+  }
+
+  @Override
+  public MuleEvent getParent() {
+    return parent;
+  }
+
+  public void setCorrelation(Correlation correlation) {
+    this.correlation = correlation;
+  }
+
+  @Override
+  public Correlation getCorrelation() {
+    return correlation;
+  }
+
+  @Override
+  public String getCorrelationId() {
+    return getCorrelation().getId().orElse(getExecutionContext().getSourceCorrelationId().orElse(getMessage().getUniqueId()));
```

Comment:
Event shouldn't care about context.id vs context.correlationId.  This logic should take into account parent event shouldn't it?